MariaDB/server
# MDEV-33545: Map innodb_flush_method=O_DIRECT_NO_FSYNC to a new innodb_doublewrite setting

- [x] *The Jira issue number for this PR is: MDEV-33545*
## Description
In commit 24648768b443f6adeb8a0f4302958bfb300d536f (MDEV-30136) the parameter `innodb_flush_method` was deprecated, with no direct replacement for `innodb_flush_method=O_DIRECT_NO_FSYNC`.

Let us change `innodb_doublewrite` from Boolean to `ENUM`, with 3 new values:

* `OFF`: Assume that writes of innodb_page_size are atomic
* `ON`: Prevent torn writes (the default)
* `fast`: Like `ON`, but avoid `fsync()` or `fdatasync()` on data files

The deprecated start-up parameter `innodb_flush_method=NO_FSYNC` will cause `innodb_doublewrite=ON` to be changed to `innodb_doublewrite=fast`, which will prevent InnoDB from making any durable writes to data files other than the doublewrite buffer. This would normally be done right before the log checkpoint LSN is updated. Depending on the file systems being used and their configuration, this may or may not be safe.

The value `innodb_doublewrite=fast` differs from the previous combination of `innodb_doublewrite=ON` and `innodb_flush_method=O_DIRECT_NO_FSYNC` by always invoking `os_file_flush_func()` on the doublewrite buffer itself in `buf_dblwr_t::flush_buffered_writes_completed()`. This should be safer when there are multiple doublewrite batches between checkpoints. Typically, once per second, `buf_flush_page_cleaner()` would write out up to `innodb_io_capacity` pages and advance the log checkpoint. Also typically, `innodb_io_capacity>128`, which is the size of the doublewrite buffer in pages. Should `os_file_flush_func()` not be invoked between doublewrite batches, writes could be reordered in an unsafe way.

The setting `innodb_doublewrite=fast` could be safe when the doublewrite buffer (the first file of the system tablespace) and the data files reside in the same file system.

Note: Enabling the setting `debug_no_sync` (in the code, `my_disable_sync`) would also disable durable writes to the log file, which would be much less safe.

Thanks to @mdcallag for reporting this.
## How can this PR be tested?
```sh
./mtr --rr innodb.alter_kill
rr replay var/log/mysqld.1.rr/mariadbd-0
rr replay var/log/mysqld.1.rr/mariadbd-1
```
and `awatch fil_system.write_through` to check that no data files are being created with `O_DSYNC` and no `fsync` or `fdatasync`  is being invoked on them.

On the first server startup of `innodb.alter_kill`, with `innodb_doublewrite=fast`, `os_file_flush_func()` would only be invoked on the `ibdata1` and possibly `ib_logfile0` (unless using a memory-mapped log). On subsequent startups with `innodb_doublewrite=OFF`, `os_file_flush_func()` will be invoked on data files during `log_checkpoint()`.

Side note: `innodb_data_file_write_through=ON` will apply also to the InnoDB temporary tablespace (`ibtmp1`) even though this is not necessary.

`IORequest::Type`: Introduce special values `WRITE_DBL` and `PUNCH_DBL` for asynchronous writes that are submitted via the doublewrite buffer. In this way, `fil_space_t::use_doublewrite()` or `buf_dblwr.in_use()` will only be consulted during `buf_page_t::flush()` and the doublewrite buffer can be enabled or disabled without any fear of inconsistency.

`buf_dblwr_t::block_size`: Replaces `block_size()`.

`buf_dblwr_t::flush_buffered_writes()`: If `!in_use()` and the doublewrite buffer is empty, just invoke `fil_flush_file_spaces()` and return. The doublewrite buffer could have been disabled while a batch was in progress.

## Basing the PR against the correct MariaDB version

## PR quality check
